SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoopInvariantCodeMotion.swift

```diff
@@ -0,0 +1,1220 @@
+         !movableInstructions.loadAndStoreAddrs.contains(accessPath),
+         movableInstructions.splitLoads(
+           analyzedInstructions: &analyzedInstructions,
+           storeAddr: storeInst.destination,
+           accessPath: accessPath,
+           context: context
+         ) {
+        movableInstructions.loadAndStoreAddrs.append(accessPath)
+      }
+    }
+  }
+
+  if !analyzedInstructions.fixLifetimes.isEmpty {
+    let sideEffectsMayRelease = analyzedInstructions.loopSideEffects.contains(where: { $0.mayRelease })
+
+    for fixLifetime in analyzedInstructions.fixLifetimes {
+      guard fixLifetime.operand.value.type.isAddress else { continue }
+
+      if sideEffectsMayRelease || !fixLifetime.mayWriteTo(
+        sideEffects: analyzedInstructions.loopSideEffects,
+        aliasAnalysis: context.aliasAnalysis
+      ) {
+        movableInstructions.sinkDown.append(fixLifetime)
+      }
+    }
+  }
+
+  for beginAccessInst in analyzedInstructions.beginAccesses {
+    if beginAccessInst.canBeHoisted(
+      outOf: loop,
+      analyzedInstructions: analyzedInstructions,
+      aliasAnalysis: context.aliasAnalysis,
+      domTree: context.dominatorTree
+    ) {
+      movableInstructions.scopedInsts.append(beginAccessInst)
+    }
+  }
+
+  return movableInstructions
+}
+
+private func optimizeLoop(
+  loop: Loop,
+  movableInstructions: inout MovableInstructions,
+  context: FunctionPassContext
+) -> Bool {
+  var changed = false
+  
+  // TODO: If we hoist tuple_element_addr and struct_element_addr instructions here, hoistAndSinkLoadsAndStores could converge after just one execution!
+  changed = movableInstructions.simpleHoistInstructions(
+      loop: loop,
+      context: context
+    ) || changed
+  
+  changed = movableInstructions.hoistAndSinkLoadsAndStores(
+    outOf: loop,
+    context: context
+  ) || changed
+
+  changed = movableInstructions.hoistInstructions(
+      loop: loop,
+      context: context
+    ) || changed
+
+  changed = movableInstructions.sinkInstructions(
+      loop: loop,
+      context: context
+    ) || changed
+
+  changed = movableInstructions.hoistWithSinkScopedInstructions(
+      loop: loop,
+      context: context
+    ) || changed
+
+  return changed
+}
+
+extension MovableInstructions {
+  mutating func splitLoads(
+    analyzedInstructions: inout AnalyzedInstructions,
+    storeAddr: Value,
+    accessPath: AccessPath,
+    context: FunctionPassContext
+  ) -> Bool {
+    var tmpLoads = Stack<LoadInst>(context)
+    defer {
+      analyzedInstructions.append(newLoads: tmpLoads)
+      tmpLoads.deinitialize()
+    }
+    var splitCounter = 0
+
+    while let loadInst = analyzedInstructions.popLoad() {
+      guard splitCounter <= 6 else {
+        tmpLoads.push(loadInst)
+        return false
+      }
+
+      guard !loadInst.isDeleted,
+        loadInst.mayRead(fromAddress: storeAddr, context.aliasAnalysis),
+        !accessPath.isEqualOrContains(loadInst.operand.value.accessPath)
```

Comment:
I think it should be the other way around:
```
   loadInst.operand.value.accessPath. accessPath.contains(accessPath)
```
(note the missing `!`)

... where `contains` excludes the "equal". You can add this to `SmallProjectionPath` by checking if the returned path from `getProjection` is not empty.

Then you also don't need the `loadInst.mayRead`, because if the store is a sub-projection of the load, the load definitely reads from the store-address.